File: config/config-sbc2ha/src/main/java/com/dfi/sbc2ha/config/sbc2ha/definition/actuator/ActuatorConfig.java
package com.dfi.sbc2ha.config.sbc2ha.definition.actuator;

import com.dfi.sbc2ha.config.sbc2ha.definition.enums.OutputKindType;
import com.dfi.sbc2ha.config.sbc2ha.definition.enums.ActuatorType;
import com.dfi.sbc2ha.helper.deserializer.DurationDeserializer;
import com.dfi.sbc2ha.helper.deserializer.DurationSerializer;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonSubTypes.Type;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import lombok.Data;

import java.time.Duration;
import java.util.Optional;


@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        property = "kind",
        include = JsonTypeInfo.As.EXISTING_PROPERTY,
        defaultImpl = McpOutputConfig.class
)
@JsonSubTypes({
        @Type(value = McpOutputConfig.class, name = "mcp"),
        @Type(value = PcaOutputConfig.class, name = "pca"),
        @Type(value = GpioOutputConfig.class, name = "gpio"),
        @Type(value = GpioPwmOutputConfig.class, name = "gpiopwm"),
        @Type(value = CoverConfig.class, name = "cover")
})
@Data
public abstract class ActuatorConfig {

    OutputKindType kind;

    @JsonProperty("name")
    String name;

    @JsonProperty("output_type")
    ActuatorType outputType;

    int output;

    /**
     * Send autodiscovery message to Home Assistant.
     * Optional
     */
    @JsonProperty("show_in_ha")
    boolean showInHa = true;

    /**
     * You can enable restore_state option. It’s bit experimental.
     * It saves state of relay or cover in state.json file, which is located in same directory as your config.json.
     * If output_type is None, then this value is overwritten to False!
     * Optional
     */
    @JsonProperty("restore_state")
    boolean restoreState = false;

    @JsonDeserialize(using = DurationDeserializer.class)
    @JsonSerialize(using = DurationSerializer.class)
    @JsonProperty("momentary_turn_on")
    Duration momentaryTurnOn;

    @JsonDeserialize(using = DurationDeserializer.class)
    @JsonSerialize(using = DurationSerializer.class)
    @JsonProperty("momentary_turn_off")
    Duration momentaryTurnOff;

    @JsonIgnore
    public String getName() {
        return Optional.of(name)
                .orElse(String.valueOf(output));
    }


}

File: config/editor/src/app/definition/actuator/ActuatorConfig.ts
import {OutputKindType} from "../enums/OutputKindType";
import {ActuatorType} from "../enums/ActuatorType";
import {Duration} from "../Duration";
import {Expose, Transform} from "class-transformer";
import {fromDuration, skipDefault, toDuration} from "../AppConfig";

export abstract class ActuatorConfig {
    kind: OutputKindType;

    name: string;

    @Expose({name: "output_type"})
    outputType: ActuatorType;

    output: number;

    @Expose({name: "show_in_ha"})
    @Transform(({value}) => skipDefault(value, true), {toPlainOnly: true})
    showInHa: boolean = true;

    @Expose({name: "restore_state"})
    @Transform(({value}) => skipDefault(value, false), {toPlainOnly: true})
    restoreState: boolean = false;

    @Transform(({value}) => toDuration(value), {toClassOnly: true})
    @Transform(({value}) => fromDuration(value), {toPlainOnly: true})
    @Expose({name: "momentary_turn_on"})
    momentaryTurnOn?: Duration;

    @Transform(({value}) => toDuration(value), {toClassOnly: true})
    @Transform(({value}) => fromDuration(value), {toPlainOnly: true})
    @Expose({name: "momentary_turn_off"})
    momentaryTurnOff?: Duration;


    protected constructor(kind: OutputKindType, name: string, outputType: ActuatorType, output: number) {
        this.kind = kind;
        this.name = name;
        this.outputType = outputType;
        this.output = output;
    }
}
